processing/src/main/java/org/apache/druid/collections/MultiColumnSorter.java

```diff
@@ -0,0 +1,113 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.druid.collections;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.collect.MinMaxPriorityQueue;
+import com.google.common.collect.Ordering;
+import org.apache.druid.java.util.common.ISE;
+
+import java.util.Comparator;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Objects;
+
+public class MultiColumnSorter<T>
```

Comment:
This class is based on a priority queue. Thus, it would seem to be a "multi column merger" of previously-sorted values. Or, are we using a priority queue to sort?

Reply:
> "multi column merger"

It is indeed "multi column merger". In many places, the TOPK algorithm is often used in this way. I want to know what better multi column sorting method you can recommend?

Reply:
Sorting is a tricky beast! First, there are the memory issues. In the worst case, someone might sort all columns for all rows in a segment, which might create excessive memory use. Druid has no good solution at present, we'd just have to be aware of the issue.

For a sort algorithm, there are many. Back in Apache Drill, we found that the Hadoop sorter worked better than the stock Java sorter. MSQ introduced a disk-based "super sorter", which is overkill here.

There are probably some sorts going on in ingestion, but a quick search of the code didn't uncover anything obvious. Perhaps a message in the Druid dev Slack channel would offer some pointers.

In any event, the sorter can be wrapped in an interface, since sorts are the kind of things that can always be improved. For this use case, perhaps we can assume all rows have the same schema, and are Java objects in an array, and just use the Java sort options, along with a comparator that compares columns in key order.